QR Code Scanner

nickfujita opened the pull request:
**What problem does this PR solve?**
Currently the only way that users are able to communicate with the neon wallet is via manual text entry and copy/paste. There are some security risks involved in the form of users doing things like copy and pasting private keys in potentially unsafe places. There can also be a risk involved for users that could potentially have malware or access malicious that access their clipboard. Even the case where they copy their private key, paste it for use in the neon wallet, forget that they had it in their clipboard, and accidentally paste and send it somewhere they didn't intend to.

**How did you solve this problem?**
By introducing QR code scanning as a new input method, we can help users to reduce the contract points where their private could be leaked. For example, they may securely print out the wallet details at the time of wallet creation instead of taking a screen shot or copying and pasting to a notepad, all on their local computer. The user may then keep this paper wallet for use with the QR code functionality. This will also help to save users time for those that are entering it manually.

Updates to new wallet creation details screen have been made to better facilitate the usefulness of this function.

This will also help to increase the ease integration points between mobile wallets, such as the [O3 Wallet](https://o3.network/).

**How did you make sure your solution works?**
I've gone ahead and added the scanning functionality to the `Login using private key` screen. In order to test this out for yourself, please navigate to that screen, click the `Scan` button. A preview of your camera will appear, and you can present a QR code of a private key to be scanned. The value from the scan upon success will call the method to login with WIF based on the value found, and attempt the login.

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+import Button from '../Button'
+import Instascan from 'instascan'
+
+import qrCodeScannerStyles from './QrCodeScanner.scss'
```
Could you use `import styles from ...` (to be consistent with other components)

nickfujita replied:
updated

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+import qrCodeScannerStyles from './QrCodeScanner.scss'
+
+type Props = {
+  callback: Function,
```
Could you type the callback?
`callback: (content: Object) => any` (I'm unsure what is the type of content)

nickfujita replied:
updated

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+}
+
+class QrCodeScanner extends Component<Props, State> {
+  scanPreviewElement: HTMLVideoElement | null
```
I think you could probably use `?HTMLVideoElement`

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+  scanPreviewElement: HTMLVideoElement | null
+
+  state = {
+    scanner: {}
```
I think `scannerInstance` is a better name

nickfujita replied:
updated

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+
+    const { callback } = this.props
+    const scannerInstance = new Instascan.Scanner({ video: this.scanPreviewElement })
+    scannerInstance.addListener('scan', function (content) {
```
Could you use arrow functions instead?

nickfujita replied:
updated

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+      scannerInstance.stop()
+      callback(content)
+    })
+    Instascan.Camera.getCameras().then(function (cameras) {
```
Same here (arrow function)

nickfujita replied:
updated

evgenyboxer commented on app/components/QrCodeScanner/QrCodeScanner.jsx:
```diff
@@ -0,0 +1,66 @@
+      callback(content)
+    })
+    Instascan.Camera.getCameras().then(function (cameras) {
+      if (cameras.length > 0) {
```
cameras will always be an array? if so, its better to type it

Sqeeki55 replied:
Mention@evgenyboxer 🏁🚩☺

evgenyboxer commented:
@nickfujita could you update the test snapshots (they are failing the ci)

nickfujita commented:
@evgenyboxer Thanks for the timely review :)

I've updated the points you requested, and added the new component to the login snapshot for automated testing.

nickfujita commented:
@comountainclimber Thanks for the heads up on this. What would you suggest as the best course of action? Once this branch is merged into the current master, should I file a new PR to merge the same into `design-v2`? Same goes for any new PRs? Should `design-v2` simply be rebased against master on a regular basis?

comountainclimber commented:
@nickfujita Im pretty sure the best course of action would be to file a separate PR for this feature => `design-v2`. It will require you to do some additional work as v2 is substantially different both in terms of UX as well as container/component structures. @mhuggins maybe you want to weigh in?

Also nice job on this!

nickfujita commented:
@comountainclimber Thanks. I'll look to make separate PRs for each of these changes to the `design-v2` branch as well

nickfujita commented:
@evgenyboxer Wanted to followup an your re-review and approval :)

comountainclimber commented:
Nice work @nickfujita let me know if you need help getting this into `design-v2`

nickfujita commented:
@comountainclimber 

Fixed. Things got a little wacky when I was adjusting for flow tests.

comountainclimber commented:
Thanks @nickfujita I will pull the latest and test 👍

drptbl commented:
@nickfujita Thank you.

Tested on Windows 10 and found some issues:
1. Unnecessary scroll bar on the bottom on "Login using a private key" page
2. "Enter your private key here" input field is stretched
3. Camera preview is misaligned after clicking on "Scan" button

I think that maybe it would be a good idea to mention somewhere that this is QR code scanner feature (so users know what they are able to scan). Currently "Scan" don't say much for users.

![scan](https://user-images.githubusercontent.com/8177587/41072097-01aaf0a4-69fc-11e8-9cf0-22a476674e6f.gif)

nickfujita commented:
- Unnecessary scroll bar on the bottom on "Login using a private key" page
_Was not seeing this on osx, as the scroll bar was on autohide. thanks for catching_

- "Enter your private key here" input field is stretched
_fixed_

- Camera preview is misaligned after clicking on "Scan" button
_could you please clarify on `misaligned`? it was intended to hang off the scan/cancel button_

- If not a valid private key QR code is scanned, button label doesn't change from "Cancel" to "Scan".(Click on "Scan", scan invalid private key QR code => error show up, camera turns off, but label didn't change to "Scan")
_Updated the scanner flow to simply add contents of qr code to pass input field instead of trying to auto login. Validation can be handled there. Fixed scan button test and changing on scan_

drptbl commented:
@nickfujita By misaligned I mean that camera preview show up in a weird place - right under the "Scan" button, there is no spacing, preview is not centered. It doesn't look good.

drptbl commented:
All right, so just re-tested it:
1. Issue I've described with misaligned camera preview is still there. I think that it should be fixed. There should be a spacing between "Scan QR code" button and camera preview. Preview should also be centered.
![center](https://user-images.githubusercontent.com/8177587/41101841-98250e02-6a65-11e8-8e2d-fdbe159e464a.PNG)

2. I liked the auto-login thing. Is there any specific reason you've removed it?

nickfujita commented:
Added back the login on scan feature. Originally took it out to simplify the error checking on valid priv key value, but after looking at it again, seems to work fine with the notification on fail.

The preview is meant to be more of a dropdown extension of the scan button, to keep user focus in that area.

comountainclimber commented:
@nickfujita can we center the camera view in the login container like @drptbl had suggested? After that I will give 👍 nice work!
<img width="1052" alt="screen shot 2018-06-10 at 9 38 01 am" src="https://user-images.githubusercontent.com/13072035/41203281-071ffece-6c92-11e8-94a9-7adb0250830f.png">

evgenyboxer commented:
@nickfujita The UI needs improvement, perhaps speak with Ben?

nickfujita commented:
@comountainclimber @drptbl  Moved scanner preview to center.

drptbl commented:
@evgenyboxer Tested, looks good!

evgenyboxer commented:
@nickfujita Great, thanks. Could you please resolve conflicts?

nickfujita commented:
@evgenyboxer resolved.

drptbl commented:
@nickfujita Can you also take a look at e2e tests for creation of new wallet, please? They seem to fail after your changes.
https://circleci.com/gh/CityOfZion/neon-wallet/2345
